### ISSUE TITLE
RFO: name for canonical file containing gene annotations

### ISSUE THREAD

**Issue by adf-ncgr** [LegumeFederation/datastore#18]
By canonical, I mean something like: 
"the file which downstream processes such as
 - chado-loading 
 - browser-building
 - mine-ification 
 - CoGe-population
should start with when a file of gene annotations is needed"

I have been slow to settle on a pattern, but it's time to make a decision, and stick with it. The pigeonpea file I just put in for #14 was "cajca.ICPL87119.gnm1.ann1.AHRD.gff3.gz" to reflect the fact that it has been run through AHRD and to avoid any dependence on the original file naming scheme which will surely lead to diversity. If we'd prefer to avoid a commitment to AHRD we could just use something generic but explicit like "cajca.ICPL87119.gnm1.ann1.canonical.gff3.gz"

**Comment by vivekkrish** [comment]
Hi @adf-ncgr, I've also been relying on the pattern suggested by you to name files that I pre-processed for A17 and R108. For example:
```
medtr.A17_HM341.gnm4.ann2.genomic.gff3.gz
medtr.R108_HM340.gnm1.ann1.genomic.gff3.gz
```

I would definitely vote for the latter option, i.e. avoid commitment to a specific method such as **AHRD** in the file name. Instead, something like **canonical** or **genomic**, might be more suited, because these files can constitute multiple types of genomic features: protein_coding, te_related, non_coding, etc.

Thoughts?

**Comment by adf-ncgr** [comment]
yes, I'm also inclined to go with a more generic option; I think I prefer **canonical** slightly over **genomic** since I guess the fact that we're talking about **gnmX.annY** already implies that it is 
**genomic**, and **canonical** or maybe **recommended** seems to imply **USE THIS ONE**, but I don't feel very strongly about it.

**Comment by vivekkrish** [comment]
Agree with **canonical** as well.

**Comment by vivekkrish** [comment]
Slightly OT: I have a question related to nomenclature (probably @cann0010 would be the best person to answer this), particularly regarding the  **key** (e.g. SWBf, cxSM, etc.) from the registry that is being used to tag a specific directory (such as _gnm1_, _gnm1.ann1_, _gnm1.syn1_, etc.), and subsequently the files within that directory:

Is the **key** being embedded in the ID prefixes within the FASTA, GFF and/or other format files?

**Comment by adf-ncgr** [comment]
It's on topic enough, but could be pulled into a separate issue if debate becomes heated enough ;)

I have not been including these keys in the embedded IDs, though if there is an argument to be made for doing so I am open to it. I continue to be slightly unsure what we are gaining from the use of the keys. Will note that some folders are inconsistent in not including them in the file names, e.g.:
https://legumeinfo.org/data/public/Cajanus_cajan/ICPL87119.gnm1.syn1.Nkx0/
though this inconsistency may be limited to the synteny folders, which I think are still awaiting being brought into full compliance, e.g. as per #16 they ought to be gff3 not gff

thanks for raising it!

**Comment by vivekkrish** [comment]
Hi @adf-ncgr, thanks for the response! My understanding regarding the **key** was that it acts as some sort of "stable" identifier (vis-a-vis DOI or ARK). It's reasonable and makes sense to tag directories/filenames with this **key**, but it is probably over-kill to tag each feature identifier as well.

Will wait for @cann0010 to chime in.

**Comment by cann0010** [comment]
@adf-ncgr, @vivekkrish - I agree that "It's reasonable and makes sense to tag directories/filenames with this key, but it is probably over-kill to tag each feature identifier as well."
I would say that the main purposes of the key are to concisely associate a set of files with their metadata, where other information such as provenance is stored - and to put our "stamp" the file set so that we can indicate any changes made with respect to other instances of these files. 

Similar keys are in use at Phytozome, with their data-set number (e.g. Gmax_275_v2.0.fa.gz) and in the thousand plant transcriptomes project, http://www.onekp.com/public_data.html (e.g. RKFX-SOAPdenovo-Trans-assembly.fa.bz2). In the case of the onekp project, the key names have been applied to the IDs within files (e.g. >scaffold-TVSH-2000001-Bituminaria_bituminosa), but they are not used by Phytozome. In our case, I think adding keys to all primary feature IDs would be too cumbersome, and I am not yet aware of a need. 

More about our data store rationales and protocols here: https://github.com/LegumeFederation/datastore/blob/master/metadata/about_the_data_repository.md (It's probably time for us to revisit and revise this - at least noting the two new in the README / metadata: moving the MD5 out of the README and adding a Changes field.

**Comment by adf-ncgr** [comment]
@cann0010 thanks for the clarification, I think it actually helps put a fine point on my main source of confusion with the keys. In your example of phytozome, is the data set number the "275" or the "v2.0" or both? In our scheme we have both the versioning component (e.g. phavu.G19833.gnm2.ann1 and the "key"=PB8d). Is the intent behind having both schemes to give us the freedom to change the versioning component if we see fit to do so? or do we just see these two things as being separate but equally unique and non-changing? What I'm trying to imagine is the scenario in which the opaque key gives us something that we don't already have with the less opaque versioning scheme. I can easily believe that we will decide there are things about the non-opaque versioning scheme that we aren't happy about, so maybe this is the answer to my nagging doubt? if this seems like a reasonable way of describing it, then I can try to document in the master file along with some of the other resolutions recently made. I had started this, but got distracted- but that's what concurrent versioning systems are for....

**Comment by cann0010** [comment]
@adf-ncgr - in the Phytozome example, "275" is the data set name. A new version (e.g. v3.0) would require a new number.

"Is the intent behind having both schemes to give us the freedom to change the versioning component if we see fit to do so?" -- NO. The key is used in parallel with the human-readable portion of the directory/file name.

Could we get by with just a human-readable naming convention? Yes - but I think it would be more prone to inconsistencies. The four-letter key is a convenience, and is pretty easy to specify. That's all.

**Comment by adf-ncgr** [comment]
@cann0010 @vivekkrish 
we've wandered a bit from the original purpose of this RFO, but I appreciate the thoughtful engagement. Getting back to the original question of "name for canonical file containing gene annotations", and taking into account the parallel discussion at #2, how about if we settle on:
gensp.Accession.gnm#.ann#.KEYX.main_genome_annotation.gff3.gz
e.g.
vigun.IT97K-499-35.gnm1.ann1.main_genome_annotation.gff3.gz

this is somewhat similar to what @vivekkrish had suggested early on, and if we don't use canonical for genomes, it seems silly to use it for their annotations.
I'll give you over the long thanksgiving weekend to formulate objections though I anticipate we're all good and tired of being filenaming-legislators ;)

**Comment by cann0010** [comment]
@adf-ncgr I have to say I'm not enthusiastic about adding another 22 characters to most of the annotation files. And the "annotation" part is redundant with ann#. 

It's also not clear to me which files would be labeled as "main_genome" and which wouldn't. Taking the phavu gnm2 files as an example, I think all of the files except the three .txt files are annotations of "main_genome"?

Can point to an example where some annotation files are not "main_genome"? 

Thinking aloud ... could the README be used to indicate "main" and "not-main" files?

  phavu.G19833.gnm2.ann1.PB8d.cds.fna.gz
  phavu.G19833.gnm2.ann1.PB8d.cds_primaryTranscript.fna.gz
  phavu.G19833.gnm2.ann1.PB8d.gene.gff3.gz
  phavu.G19833.gnm2.ann1.PB8d.gene_exons.gff3.gz
  phavu.G19833.gnm2.ann1.PB8d.locus_transcript_name_map.txt
  phavu.G19833.gnm2.ann1.PB8d.protein.faa.gz
  phavu.G19833.gnm2.ann1.PB8d.protein_primaryTranscript.faa.gz
  phavu.G19833.gnm2.ann1.PB8d.repeatmasked_assembly_v2.0.gff3.gz
  phavu.G19833.gnm2.ann1.PB8d.synonym.txt
  phavu.G19833.gnm2.ann1.PB8d_info.txt
  phavu.G19833.gnm2.ann1.PB8d.transcript.fna.gz
  phavu.G19833.gnm2.ann1.PB8d.transcript_primaryTranscript.fna.gz

**Comment by adf-ncgr** [comment]
@cann0010 to be clear I am only talking now about one file (a gff3 file) and it will probably in no case be one of the files that is in the original set because it will have our "federation-approved" IDs and possibly also enhanced content like AHRD-based functional descriptors (though there are good arguments for leaving those out of the the gff3 file as well). My main intent is not to supplant the original files but to indicate to others what we have used as the basis for populating our downstream systems (with the intent that if they do the same, federation-magic will require fewer regexp-based spells for linking with names/IDs). 

So, in your phavu example, it would actually be none of those files which I think are direct from
phytozome (modulo the prepending of phavu.G19833.gnm2.ann1.PB8d.), but instead the file
that I shared with Wei, but I guess never made it's way to the data store proper (probably because I haven't actually loaded into chado, pending outstanding questions about how to deal with upgrading to new versions in the system as a whole).

Your example is actually a good one, in that most people will be horribly confused as to why there is both:
phavu.G19833.gnm2.ann1.PB8d.gene.gff3.gz
phavu.G19833.gnm2.ann1.PB8d.gene_exons.gff3.gz

of course, it may not help to provide yet another gff3 as a choice but my hope is that if we have
a consistent convention for naming one as our generally recommended choice it will lead to less confusion.

regarding your comment the
> the "annotation" part is redundant with ann#.
I agree, but the same is true for gnm and "main_genome"; I was just trying to be somewhat consistent...

let me know your further thoughts- definitely not trying to impose my will, no matter how benevolent my motives...

**Comment by cann0010** [comment]
I see. That's more palatable. Precisely what the name should be: am still thinking. Adding "main_" or "_main" could suffice - as in "gene_models_main.gff3" to distinguish from special-purpose files.

Another big step will be to establish preferred names and suffixes for all file types. We have quite a variety of forms now, alas. Am reviewing them all now; will try to make a proposal next week.

**Comment by adf-ncgr** [comment]
gene_models_main.gff3 has been established- closing
